README
NAME
    Business::OnlinePayment::CyberSource - CyberSource backend for
    Business::OnlinePayment

VERSION
    version 3.000011

SYNOPSIS
            use Business::OnlinePayment;

            my $tx = Business::OnlinePayment->new( "CyberSource" );
            $tx->content(
                    login          => 'username',
                    password       => 'password',
                    type           => 'CC',
                    action         => 'Normal Authorization',
                    invoice_number => '00000001', # MurchantReferenceCode
                    first_name     => 'Peter',
                    last_name      => 'Bowen',
                    address        => '123 Anystreet',
                    city           => 'Orem',
                    state          => 'Utah',
                    zip            => '84097',
                    country        => 'US',
                    email          => 'foo@bar.net',
                    card_number    => '4111111111111111',
                    expiration     => '09/06',
                    cvv2           => '1234', #optional
                    amount         => '5.00',
                    currency       => 'USD',
            );

            $tx->submit();

            if($tx->is_success()) {
                    print "Card processed successfully: ".$tx->authorization."\n";
            } else {
                    print "Card was rejected: ".$tx->error_message."\n";
            }

            ####
            # Two step transaction, authorization and capture.
            # If you don't need to review order before capture, you can
            # process in one step as above.
            ####

      $tx = Business::OnlinePayment->new("CyberSource");
            $tx->content(
                    login          => 'username',
                    password       => 'password',
                    type           => 'CC',
                    action         => 'Authorization Only',
                    invoice_number  => 44544, # MurchantReferenceCode
                    description     => 'Business::OnlinePayment visa test',
                    amount          => '42.39',
                    first_name      => 'Tofu',
                    last_name       => 'Beast',
                    address         => '123 Anystreet',
                    city            => 'Anywhere',
                    state           => 'Utah',
                    zip             => '84058',
                    country         => 'US',
                    email           => 'tofu@beast.org',
                    card_number     => '4111111111111111',
                    expiration      => '12/25',
                    cvv2            => 1111,
            );
            $tx->submit();

            if($tx->is_success()) {
                    # get information about authorization
                    my $authorization = $tx->authorization();
                    my $order_number = $tx->order_number(); # RequestId
                    my $avs_code = $tx->avs_code(); # AVS Response Code();
                    my $cvv2_response = $tx->cvv2_response(); # CVV2/CVC2/CID Response Code();

                    # now capture transaction

                    $tx->content(
                            login          => 'username',
                            password       => 'password',
                            type           => 'CC',
                            action         => 'Post Authorization',
                            invoice_number => 44544, #MurchantReferenceCode
                            amount         => '42.39',
                            po_number       => $tx->order_number(), # RequestId
                    );

                    $tx->submit();

                    if($tx->is_success()) {
                            print "Funds captured successfully\n";
                    } else {
                            print "Card was rejected: ".$tx->error_message."\n";
                    }

            } else {
                    print "Card was rejected: " . $tx->error_message() . "\n";
            }

DESCRIPTION
    For detailed information see Business::OnlinePayment.

METHODS
  BUILD
    this is a before-construction hook for Moose. You Will never call this
    method directly.

SUPPORTED TRANSACTION TYPES
  CC
    Content required: type, login, action, amount, first_name, last_name,
    card_number, expiration.

    cvv2 is required in order to get back a cvv2_response value.

  Settling
    To settle an authorization-only transaction (where you set action to
    "Authorization Only"), submit the "order_number" code in the field
    "po_number" with the action set to "Post Authorization".

    You can get the transaction id from the authorization by calling the
    "order_number" method on the object returned from the authorization. You
    must also submit the amount field with a value less than or equal to the
    amount specified in the original authorization.

ACKNOWLEDGMENTS
    Jason Kohles
        For writing BOP - I didn't have to create my own framework.

    Ivan Kohler
        Tested the first pre-release version and fixed a number of bugs. He
        also encouraged me to add better error reporting for system errors.
        He also added failure_status support.

    Jason (Jayce^) Hall
        Adding Request Token Requirements (Among other significant
        improvements... )

SEE ALSO
    Business::OnlinePayment

BUGS
    Please report any bugs or feature requests on the bugtracker website
    https://github.com/hostgator/Business-OnlinePayment-CyberSource/issues

    When submitting a bug or request, please include a test-file or a patch
    to an existing test-file that illustrates the bug or desired feature.

AUTHORS
    *   Jad Wauthier <Jadrien dot Wauthier at GMail dot com>

    *   Caleb Cushing <xenoterracide@gmail.com>

    *   Peter Bowen <peter@bowenfamily.org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2012 by HostGator.com
    <http://www.hostgator.com>.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for Business-OnlinePayment-CyberSource

3.000011  2012-11-15
	- Change minimum version of B::C to 0.007006
	- Update MetaResources

3.000010  2012-10-29
	- Removed checking of accept role
	- removed debugging output

3.000009  2012-10-15
	- Remove unneeded git remote

3.000008  2012-10-15
	- Updated logic in Client class to rethrow exceptions thrown by the CyberSource response object

3.000007  2012-10-04
	- improve conditional checks for avs, authorization and cvcode

3.000006  2012-10-04
	- Add conditionals to only set avs_code and authorization if the appropriate roles are applied to the response object
	- Fix some logic errors in client object
	- Clean up some logic to remove unnecessary variables

3.000005  2012-09-04
	- Specify minimum version of Business::CyberSource as 0.006000
	- Set response values for success and failure
	- Add auth failure tests

3.000004  2012-08-30
	- fix dependencies

3.000003  2012-08-28
	- fix licensing
	- release to cpan

3.000002  2012-08-24
	- Fixed logic error preventing follow-on credits from working properly
	- Fixed Logic errors with expiration dates
	- Added support for expiration dates with MMYY format
	- various other bug fixes

3.000000  2012-08-16
	- Refactored code to use Business::CyberSource as the underlying engine

2.01      2011-07-25
	- use dist-zilla
	- mostly bring up to perl-critic compliance
	- allow passing a config hash on object initialization
	- fix to work with CyberSource Simple API 5.0
	- new maintainership Caleb Cushing

2.00	2007-04-21
	- release of 2.00, no changes noted, and I'm way to lazy to look at diff's

0.09	2007-03-01
	- Updated by Jason Hall (Jayce^) <jayce+cpan@lug-nut.com>
	- Updated to New Cybersource API
	- Specifically using the new 'requestToken' field that is necessary
	- Internal BOP option of 'security_key' to normalize its usage

0.08   2005-12-31
	- date unrecorded, so I made one up : Caleb Cushing
	- This update by Ivan Kohler <ivan-cybersource@420.am>
	- Add my previous patch against 0.6 to fix an explicitly specified
	  conf_file
	- Add "failure statuses" with a B:OP pre-3.00_04 workaround

0.07	2005-12-12
	- Improved the Documentation

0.06	2005-10-12
	- Allow both line items and grand total.  If you use both, the grand total will prevail.

0.05	2005-10-02
	- Set the defaults for the business logic settings using the cybs.ini

0.04	2005-08-09
	- Found a bug that allowed the system to point to production even if the cybs.conf said to use dev

0.03	2005-06-16
	- Ivan tested this and it was pretty broken.
	- Fixed the regexp for expiration date so that it works (Ivan patch)
		- Modified it slightly to allow for dates w/o slashes
	- Allowed for custom cybs.ini locations (Ivan Patch)
		- Made the default /etc/ as opposed to ./
	- Added a ton of error handling after Ivan had cats over the poor error
	  handling before
	  (Peter is being generous.  I was downright bitchy. -Ivan)

0.02	2005-01-31
	- First Release

0.01	2005-01-26
	- original version; created by hand.


Makefile.PL

use strict;
use warnings;

use 5.010;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "CyberSource backend for Business::OnlinePayment",
  "AUTHOR" => "Jad Wauthier <Jadrien dot Wauthier at GMail dot com>, Caleb Cushing <xenoterracide\@gmail.com>, Peter Bowen <peter\@bowenfamily.org>",
  "BUILD_REQUIRES" => {
    "File::Find" => 0,
    "File::Temp" => 0,
    "Test::More" => 0
  },
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "Business-OnlinePayment-CyberSource",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "Business::OnlinePayment::CyberSource",
  "PREREQ_PM" => {
    "Business::CyberSource" => "0.006000",
    "Business::CyberSource::Client" => "0.007006",
    "Business::OnlinePayment" => 0,
    "Class::Load" => "0.20",
    "Data::Dump" => 0,
    "DateTime" => 0,
    "Exception::Base" => 0,
    "Moose" => 0,
    "Moose::Role" => 0,
    "MooseX::Aliases" => 0,
    "MooseX::NonMoose" => 0,
    "MooseX::StrictConstructor" => 0,
    "MooseX::Types::Common::String" => 0,
    "MooseX::Types::CyberSource" => 0,
    "MooseX::Types::Moose" => 0,
    "Try::Tiny" => 0,
    "namespace::autoclean" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "VERSION" => "3.000011",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);





lib/Business/OnlinePayment/CyberSource/Error.pm
package Business::OnlinePayment::CyberSource::Error;
use 5.006;
use strict;
use warnings;

BEGIN {
	our $VERSION = '2.02'; # VERSION
}
use parent 'Exporter';
use AutoLoader;

#ERROR MAP
my $error_codes = {
	'100' => {
		'Text'   => 'Successful transaction.',
		'Action' => ''
	},
	'101' => {
		'Text' => 'The request is missing one or more required fields.',
		'Action' =>
			'See the reply fields missingField_0...N for which fields are missing. Resend the request with the complete information.'
	},
	'102' => {
		'Text' => 'One or more fields in the request contains invalid data.',
		'Action' =>
			'See the reply fields invalidField_0...N for which fields are invalid. Resend the request with the correct information.'
	},
	'150' => {
		'Text'   => 'Error: General system failure.',
		'Action' => 'Wait a few minutes and resend the request.'
	},
	'151' => {
		'Text' =>
			'Error: The request was received but there was a server timeout. This error does not include timeouts between the client and the server.',
		'Action' =>
			'To avoid duplicating the transaction, do not resend the request until you have reviewed the transaction status in the Support Screens.'
	},
	'152' => {
		'Text' =>
			'Error: The request was received, but a service did not finish running in time.',
		'Action' =>
			'To avoid duplicating the transaction, do not resend the request until you have reviewed the transaction status in the Support Screens.'
	},
	'200' => {
		'Text' =>
			'The authorization request was approved by the issuing bank but declined by CyberSource because it did not pass the Address Verification Service (AVS) check.',
		'Action' =>
			'You can capture the authorization, but consider reviewing the order for the possibility of fraud.'
	},
	'201' => {
		'Text' =>
			'The issuing bank has questions about the request. You do not receive an authorization code programmatically, but you might receive one verbally by calling the processor.',
		'Action' =>
			'Call your processor to possibly receive a verbal authorization. For contact phone numbers, refer to your merchant bank information.'
	},
	'202' => {
		'Text'   => 'Expired card.',
		'Action' => 'Request a different card or other form of payment.',
		'Status' => 'expired',
	},
	'203' => {
		'Text' =>
			'General decline of the card. No other information provided by the issuing bank.',
		'Action' => 'Request a different card or other form of payment.',
		'Status' => 'decline',
	},
	'204' => {
		'Text'   => 'Insufficient funds in the account.',
		'Action' => 'Request a different card or other form of payment.',
		'Status' => 'nsf',
	},
	'205' => {
		'Text' => 'Stolen or lost card.',
		'Action' =>
			'Refer the transaction to your customer support center for manual review.',
		'Status' => 'stolen',
	},
	'207' => {
		'Text'   => 'Issuing bank unavailable.',
		'Action' => 'Wait a few minutes and resend the request.'
	},
	'208' => {
		'Text' =>
			'Inactive card or card not authorized for card-not-present transactions.',
		'Action' => 'Request a different card or other form of payment.',
		'Status' => 'inactive',
	},
	'209' => {
		'Text' =>
			'American Express Card Identification Digits (CID) did not match.',
		'Action' => 'Request a different card or other form of payment.'
	},
	'210' => {
		'Text'   => 'The card has reached the credit limit.',
		'Action' => 'Request a different card or other form of payment.',
		'Status' => 'nsf',
	},
	'211' => {
		'Text'   => 'Invalid card verification number.',
		'Action' => 'Request a different card or other form of payment.'
	},
	'221' => {
		'Text' =>
			"The customer matched an entry on the processor's negative file.",
		'Action' => 'Review the order and contact the payment processor.',
		'Status' => 'blacklisted',
	},
	'230' => {
		'Text' =>
			'The authorization request was approved by the issuing bank but declined by CyberSource because it did not pass the card verification (CV) check.',
		'Action' =>
			'You can capture the authorization, but consider reviewing the order for the possibility of fraud.'
	},
	'231' => {
		'Text'   => 'Invalid account number.',
		'Action' => 'Request a different card or other form of payment.'
	},
	'232' => {
		'Text' => 'The card type is not accepted by the payment processor.',
		'Action' =>
			'Contact your merchant bank to confirm that your account is set up to receive the card in question.'
	},
	'233' => {
		'Text'   => 'General decline by the processor.',
		'Action' => 'Request a different card or other form of payment.',
		'Status' => 'decline',
	},
	'234' => {
		'Text' =>
			'There is a problem with your CyberSource merchant configuration.',
		'Action' =>
			'Do not resend the request. Contact Customer Support to correct the configuration problem.'
	},
	'235' => {
		'Text' =>
			'The requested amount exceeds the originally authorized amount. Occurs, for example, if you try to capture an amount larger than the original authorization amount.',
		'Action' =>
			'Issue a new authorization and capture request for the new amount.'
	},
	'236' => {
		'Text'   => 'Processor failure.',
		'Action' => 'Wait a few minutes and resend the request.'
	},
	'237' => {
		'Text'   => 'The authorization has already been reversed.',
		'Action' => 'No action required.'
	},
	'238' => {
		'Text'   => 'The authorization has already been captured.',
		'Action' => 'No action required.'
	},
	'239' => {
		'Text' =>
			'The requested transaction amount must match the previous transaction amount.',
		'Action' => 'Correct the amount and resend the request.'
	},
	'240' => {
		'Text' =>
			'The card type sent is invalid or does not correlate with the credit card number.',
		'Action' =>
			'Confirm that the card type correlates with the credit card number specified in the request, then resend the request.'
	},
	'241' => {
		'Text' => 'The request ID is invalid.',
		'Action' =>
			'Request a new authorization, and if successful, proceed with the capture.'
	},
	'242' => {
		'Text' =>
			'You requested a capture, but there is no corresponding, unused authorization record. Occurs if there was not a previously successful authorization request or if the previously successful authorization has already been used by another capture request.',
		'Action' =>
			'Request a new authorization, and if successful, proceed with the capture.'
	},
	'250' => {
		'Text' =>
			'Error: The request was received, but there was a timeout at the payment processor.',
		'Action' =>
			'To avoid duplicating the transaction, do not resend the request until you have reviewed the transaction status in the Support Screens.'
	},
	'400' => {
		'Text'   => "The Advanced Fraud Screen score exceeds your threshold.",
		'Action' => "Review the customer's order."
	},
	'510' => {
		'Text' =>
			"The authorization request was approved by the issuing bank but declined by CyberSource because it did not pass the Smart Authorization check.",
		'Action' =>
			"Do not capture the authorization without further review. The Smart Authorization codes give you additional information as to why CyberSource refused the request."
	},
	'700' => {
		'Text' =>
			"The customer is on a list issued by the U.S. government containing entities with whom trade is restricted.",
		'Action' => "Reject the customer's order.",
		'Status' => 'blacklisted',
	},
};

my $afs_codes = {
	'A' =>
		"Excessive address change. The customer had two or more billing address changes in the last six months.",
	'B' =>
		"BIN number mismatch. The customer's Visa or MasterCard credit card was issued in a country different from that specified in the billing address.",
	'C' =>
		"High count of unique account numbers. The customer used more than six unique credit cards in the last six months.",
	'D' =>
		"Domain (host) impact. The customer had a risky IP or email address.",
	'F' =>
		"Fraud list flag. The account number, street address, email address, or IP address for this order appears on the negative list for your CyberSource merchant ID.",
	'G' =>
		"Geolocation inconsistencies. The customer's email domain, phone number, billing address, shipping address, or IP address is suspicious.",
	'H' =>
		"Excessive name change. The customer had two or more name changes in the last six months.",
	'I' =>
		"Internet inconsistencies. The IP address and email domain are inconsistent with the billing address.",
	'N' =>
		"Nonsensical input. The customer name and address fields contain words or language having no meaning.",
	'O' => "Obscenities. The customer input contains obscene words.",
	'Q' => "Phone inconsistencies. The customer phone number is suspicious.",
	'R' =>
		"Risky transaction. Characteristics in this order display multiple high-risk correlations between the transaction, consumer, and merchant information.",
	'T' =>
		"Time hedge. The customer is attempting a purchase outside of the expected hours for an item.",
	'U' =>
		"Unverifiable address. The billing or shipping address cannot be verified.",
	'V' =>
		"Purchase frequency (Velocity). The account number was used many times in the past 15 minutes.",
	'W' =>
		"Warning. The billing or shipping address is similar to an address previously marked as suspect.",
	'Y' =>
		"Gift Order. The street address, city, state, or country of the billing and shipping addresses do not correlate.",
	'Z' =>
		"Invalid value. The information in the request contains an unusual or unexpected value, and a default value was substituted. Although the transaction can still be processed, examine the request carefully for abnormalities in the order."
};

my $avs_codes = {
	'A' =>
		"Street address matches, but both 5-digit and 9-digit ZIP code do not match.",
	'B' =>
		"Street address matches, but postal code not verified. Returned only for non-U.S.-issued Visa cards.",
	'C' =>
		"Street address and postal code not verified. Returned only for non-U.S.-issued Visa cards.",
	'D & M' =>
		"Street address and postal code both match. Returned only for non-U.S.-issued Visa cards.",
	'E' => "AVS data is invalid.",
	'G' => "Non-U.S. issuing bank does not support AVS.",
	'I' =>
		"Address not verified. Returned only for non-U.S.-issued Visa cards.",
	'J' =>
		"Card member's name, billing address, and postal code all match. Shipping information verified and chargeback protection guaranteed through the Fraud Protection Program.",
	'K' =>
		"Card member's name matches. Both billing address and billing postal code do not match.",
	'L' =>
		"Card member's name matches. Billing postal code matches, but billing address does not match.",
	'N' =>
		"Street address, 5-digit ZIP code, and 9-digit ZIP code all do not match.",
	'O' =>
		"Card member's name matches. Billing address matches, but billing postal code does not match.",
	'P' =>
		"Postal code matches, but street address not verified. Returned only for non-U.S.-issued Visa cards.",
	'Q' =>
		"Card member's name, billing address, and postal code all match. Shipping information verified but chargeback protection not guaranteed (Standard program).",
	'R' => "System unavailable.",
	'S' => "U.S.-issuing bank does not support AVS.",
	'U' =>
		"Address information unavailable. Returned if non-U.S. AVS is not available or if the AVS in a U.S. bank is not functioning properly.",
	'V' =>
		"Card member's name matches. Both billing address and billing postal code match.",
	'W' => "Street address does not match, but 9-digit ZIP code matches.",
	'X' => "Street address and 9-digit ZIP code both match.",
	'Y' => "Street address and 5-digit ZIP code both match.",
	'Z' => "Street address does not match, but 5-digit ZIP code matches.",
	'1' => "AVS is not supported for this processor or card type.",
	'2' => "The processor returned an unrecognized value for the AVS response.",
};

### Constructor
sub new {
	my $class = shift;
	$class = ref($class) || $class;

	my $self = {};
	bless $self, $class;

	return $self;
}

sub get_text {
	my ( $self, $error_code ) = @_;
	return $error_codes->{$error_code}->{'Text'};
}

sub get_failure_status {
	my ( $self, $error_code ) = @_;
	no warnings 'uninitialized';    ## no critic ( TestingAndDebugging::ProhibitNoWarnings )
	return $error_codes->{$error_code}->{'Status'};
}

sub get_action {
	my ( $self, $error_code ) = @_;
	return $error_codes->{$error_code}->{'Action'};
}

sub get_AVS_Text {
	my ( $self, $avs_code ) = @_;
	return $avs_codes->{$avs_code};
}

sub get_AFS_codes {
	my ( $self, $afs_code ) = @_;
	return $afs_codes ($afs_code);
}

1;

# ABSTRACT:  Error Code class for Business::OnlinePayment::CyberSource



__END__
=pod

=head1 NAME

Business::OnlinePayment::CyberSource::Error - Error Code class for Business::OnlinePayment::CyberSource

=head1 VERSION

version 2.02

=head1 SYNOPSIS

  use Business::OnlinePayment::CyberSource::Error;

  my $error_code = 100;
  my $error_decoder = new Business::OnlinePayment::CyberSource::Error;

  my $error_text     = $error_decoder->get_text($error_code);
  my $failure_status = $error_decoder->get_failure_status($error_code);
  my $error_action   = $error_decoder->get_action($error_code);

=head1 API

=over 4

=item C<get_AFS_codes>

=item C<get_AVS_Text>

=item C<get_action>

=item C<get_failure_status>

=item C<get_text>

=item C<new>

=back

=head1 SEE ALSO

perl(1). L<Business::OnlinePayment>. L<Business::OnlinePayment::CyberSource>.

=head1 BUGS

Please report any bugs or feature requests on the bugtracker website
https://github.com/xenoterracide/Business-OnlinePayment-CyberSource/issues

When submitting a bug or request, please include a test-file or a
patch to an existing test-file that illustrates the bug or desired
feature.

=head1 AUTHORS

=over 4

=item *

Peter Bowen <peter@bowenfamily.org>

=item *

Caleb Cushing <xenoterracide@gmail.com>

=back

=head1 COPYRIGHT AND LICENSE

This software is Copyright (c) 2011 by Caleb Cushing.

This is free software, licensed under:

  The Artistic License 2.0 (GPL Compatible)

=cut



lib/Business/OnlinePayment/CyberSource/Role/InputHandling.pm
package Business::OnlinePayment::CyberSource::Role::InputHandling;

use 5.010;
use strict;
use warnings;
use namespace::autoclean;

use Moose::Role;

# ABSTRACT:  Input handling convenience methods for Business::OnlinePayment::CyberSource
our $VERSION = '3.000011'; # VERSION

#### Subroutine Definitions ####

# Converts input into a hashref
# Accepts:  A hash or reference to a hash
# Returns:  A reference to the supplied hash

sub _parse_input { ## no critic ( Subroutines::ProhibitUnusedPrivateSubroutines )
	my ( undef, @args ) = @_;
	my $data            = {};

	# shift off first element if only one exists and is of type HASH
	if ( scalar @args == 1 && ref $args[0] eq 'HASH' ) {
		$data             = shift @args;
	}
	# Cast into a hash if number of elements is even and first element is a string
	elsif ( ( scalar @args % 2 ) == 0 && ref $args[0] eq '' ) {
		$data             = { @args };
	}

	return $data;
}

1;

__END__

=pod

=head1 NAME

Business::OnlinePayment::CyberSource::Role::InputHandling - Input handling convenience methods for Business::OnlinePayment::CyberSource

=head1 VERSION

version 3.000011

=head1 SYNOPSIS

  package Thing;

  use Moose;

  with 'Business::OnlinePayment::CyberSource::Role::InputHandling';

  sub blah {
  	my ( $self, @args ) = @_;
		my $data = $self->_parse_input( @args );

  	$data->{color} = 'red' unless $data->{color};
  }
	 1;

  my $thing = Thing->new();

  $thing->blah( color => 'blue' );
  $thing->blah( { color => 'blue' } );

=head1 DESCRIPTION

This role provides consumers with convenience methods for handling input.

=head1 METHODS

=head2 _parse_input

Converts input into a hashref

Accepts:  A hash or reference to a hash
Returns:  A reference to the supplied hash

=head1 BUGS

Please report any bugs or feature requests on the bugtracker website
https://github.com/hostgator/Business-OnlinePayment-CyberSource/issues

When submitting a bug or request, please include a test-file or a
patch to an existing test-file that illustrates the bug or desired
feature.

=head1 AUTHORS

=over 4

=item *

Jad Wauthier <Jadrien dot Wauthier at GMail dot com>

=item *

Caleb Cushing <xenoterracide@gmail.com>

=item *

Peter Bowen <peter@bowenfamily.org>

=back

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2012 by L<HostGator.com|http://www.hostgator.com>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/Business/OnlinePayment/CyberSource/Role/ErrorReporting.pm
package Business::OnlinePayment::CyberSource::Role::ErrorReporting;

use 5.010;
use strict;
use warnings;
use namespace::autoclean;

use Moose::Role;
use MooseX::StrictConstructor;
use MooseX::Types::Moose qw(Str);
use MooseX::Types::Common::String qw(NonEmptySimpleStr);

# ABSTRACT:  Error reporting role for BOP::CyberSource
our $VERSION = '3.000011'; # VERSION

#### Subroutine Definitions ####

#### Object Attributes ####

has error          => (
	isa       => Str,
	is        => 'rw',
	required  => 0,
	predicate => 'has_errors',
	clearer   => 'clear_error',
	reader    => 'error_message',
	writer    => 'set_error_message',
	init_arg  => undef,
	lazy      => 0,
);

has failure_status => (
	isa       => Str,
	is        => 'rw',
	required  => 0,
	predicate => 'has_failure_status',
	clearer   => 'clear_failure_status',
	init_arg  => undef,
	lazy      => 0,
);

1;

__END__

=pod

=head1 NAME

Business::OnlinePayment::CyberSource::Role::ErrorReporting - Error reporting role for BOP::CyberSource

=head1 VERSION

version 3.000011

=head1 SYNOPSIS

  package Thing;

  use Moose;

  with 'Business::OnlinePayment::CyberSource::Role::ErrorReporting';
  1;

  my $thing = Thing->new();

  if ( $thing->has_errors() ) {
	  my $errors = $thing->errors();
	}

=head1 DESCRIPTION

This role provides consumers with an errors array attribute and supporting methods.

=head1 METHODS

=head2 has_errors

=head2 has_response_status

=head2 error_message

=head1 BUGS

Please report any bugs or feature requests on the bugtracker website
https://github.com/hostgator/Business-OnlinePayment-CyberSource/issues

When submitting a bug or request, please include a test-file or a
patch to an existing test-file that illustrates the bug or desired
feature.

=head1 AUTHORS

=over 4

=item *

Jad Wauthier <Jadrien dot Wauthier at GMail dot com>

=item *

Caleb Cushing <xenoterracide@gmail.com>

=item *

Peter Bowen <peter@bowenfamily.org>

=back

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2012 by L<HostGator.com|http://www.hostgator.com>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/Business/OnlinePayment/CyberSource.pm
package Business::OnlinePayment::CyberSource;

use 5.010;
use strict;
use warnings;
use namespace::autoclean;

use Moose;
use Data::Dump 'dump';
use Exception::Base;
use MooseX::NonMoose;
use MooseX::StrictConstructor;
use MooseX::Types::Moose qw(Bool HashRef Int);
use MooseX::Types::Common::String qw(NonEmptySimpleStr);

# ABSTRACT:  CyberSource backend for Business::OnlinePayment
our $VERSION = '3.000011'; # VERSION

extends 'Business::OnlinePayment';

#### Subroutine Definitions ####

# Post-construction hook
# Accepts:  A reference to a hash of construction parameters
# Returns:  Nothing

sub BUILD {
	my ( $self ) = @_;
	my $fields   = [ qw(type action reference_code amount) ];

	$self->required_fields( @$fields );

	return;
}

#### Object Attributes ####

#### Applied Roles ####

with 'Business::OnlinePayment::CyberSource::Role::TransactionHandling';

#### Method Modifiers ####

#### Meta class stuff ####

__PACKAGE__->meta->make_immutable();

1;

__END__

=pod

=head1 NAME

Business::OnlinePayment::CyberSource - CyberSource backend for Business::OnlinePayment

=head1 VERSION

version 3.000011

=head1 SYNOPSIS

	use Business::OnlinePayment;

	my $tx = Business::OnlinePayment->new( "CyberSource" );
	$tx->content(
		login          => 'username',
		password       => 'password',
		type           => 'CC',
		action         => 'Normal Authorization',
		invoice_number => '00000001', # MurchantReferenceCode
		first_name     => 'Peter',
		last_name      => 'Bowen',
		address        => '123 Anystreet',
		city           => 'Orem',
		state          => 'Utah',
		zip            => '84097',
		country        => 'US',
		email          => 'foo@bar.net',
		card_number    => '4111111111111111',
		expiration     => '09/06',
		cvv2           => '1234', #optional
		amount         => '5.00',
		currency       => 'USD',
	);

	$tx->submit();

	if($tx->is_success()) {
		print "Card processed successfully: ".$tx->authorization."\n";
	} else {
		print "Card was rejected: ".$tx->error_message."\n";
	}

	####
	# Two step transaction, authorization and capture.
	# If you don't need to review order before capture, you can
	# process in one step as above.
	####

  $tx = Business::OnlinePayment->new("CyberSource");
	$tx->content(
		login          => 'username',
		password       => 'password',
		type           => 'CC',
		action         => 'Authorization Only',
		invoice_number  => 44544, # MurchantReferenceCode
		description     => 'Business::OnlinePayment visa test',
		amount          => '42.39',
		first_name      => 'Tofu',
		last_name       => 'Beast',
		address         => '123 Anystreet',
		city            => 'Anywhere',
		state           => 'Utah',
		zip             => '84058',
		country         => 'US',
		email           => 'tofu@beast.org',
		card_number     => '4111111111111111',
		expiration      => '12/25',
		cvv2            => 1111,
	);
	$tx->submit();

	if($tx->is_success()) {
		# get information about authorization
		my $authorization = $tx->authorization();
		my $order_number = $tx->order_number(); # RequestId
		my $avs_code = $tx->avs_code(); # AVS Response Code();
		my $cvv2_response = $tx->cvv2_response(); # CVV2/CVC2/CID Response Code();

		# now capture transaction

		$tx->content(
			login          => 'username',
			password       => 'password',
			type           => 'CC',
			action         => 'Post Authorization',
			invoice_number => 44544, #MurchantReferenceCode
			amount         => '42.39',
			po_number       => $tx->order_number(), # RequestId
		);

		$tx->submit();

		if($tx->is_success()) {
			print "Funds captured successfully\n";
		} else {
			print "Card was rejected: ".$tx->error_message."\n";
		}

	} else {
		print "Card was rejected: " . $tx->error_message() . "\n";
	}

=head1 DESCRIPTION

For detailed information see L<Business::OnlinePayment>.

=head1 METHODS

=head2 BUILD

this is a before-construction hook for Moose.  You Will never call this method directly.

=head1 SUPPORTED TRANSACTION TYPES

=head2 CC

Content required: type, login, action, amount, first_name, last_name, card_number, expiration.

cvv2 is required in order to get back a cvv2_response value.

=head2 Settling

To settle an authorization-only transaction (where you set action to
C<Authorization Only>), submit the C<order_number> code in the field
C<po_number> with the action set to C<Post Authorization>.

You can get the transaction id from the authorization by calling the
C<order_number> method on the object returned from the authorization.
You must also submit the amount field with a value less than or equal
to the amount specified in the original authorization.

=head1 ACKNOWLEDGMENTS

=over 4

=item Jason Kohles

For writing BOP - I didn't have to create my own framework.

=item Ivan Kohler

Tested the first pre-release version and fixed a number of bugs.
He also encouraged me to add better error reporting for system
errors.  He also added failure_status support.

=item Jason (Jayce^) Hall

Adding Request Token Requirements (Among other significant improvements... )

=back

=head1 SEE ALSO

L<Business::OnlinePayment>

=head1 BUGS

Please report any bugs or feature requests on the bugtracker website
https://github.com/hostgator/Business-OnlinePayment-CyberSource/issues

When submitting a bug or request, please include a test-file or a
patch to an existing test-file that illustrates the bug or desired
feature.

=head1 AUTHORS

=over 4

=item *

Jad Wauthier <Jadrien dot Wauthier at GMail dot com>

=item *

Caleb Cushing <xenoterracide@gmail.com>

=item *

Peter Bowen <peter@bowenfamily.org>

=back

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2012 by L<HostGator.com|http://www.hostgator.com>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut
